holidays/countries/tonga.py

```diff
@@ -0,0 +1,237 @@
+        # Topou VI: Jul 4 (2015-Present)**
+        # George Tupou V: Aug 1 (2008-2011)
+        # Tāufa'āhau Tupou IV: Jul 4: (1967-2006)**
+        #  *  No celebration for in-between years i.e. 2007, 2012-2014.
+        #  ** Has de facto merged with King's Birthday.
+
+        if 2008 <= self._year <= 2011:
+            aug_1 = self._add_holiday_aug_1(
+                tr(
+                    # Anniversary of the Coronation Day of the reigning Sovereign of Tonga.
+                    "Fakamanatu 'o e 'Aho Hilifaki Kalauni 'o 'Ene 'Afio ko e Tu'i 'o Tonga "
+                    "'a ia 'oku lolotonga Pule"
+                )
+            )
+            if self._year <= 2009:
+                self._add_observed(aug_1)
+            else:
+                self._add_observed(
+                    aug_1, rule=TUE_WED_TO_PREV_MON + THU_FRI_TO_NEXT_MON + SAT_SUN_TO_NEXT_MON
+                )
+
+        # Constitution Day.
+        nov_4 = self._add_holiday_nov_4(tr("'Aho Konisitutone"))
+        if self._year <= 2009:
+            self._add_observed(nov_4)
+        else:
+            self._add_observed(
+                nov_4, rule=TUE_WED_TO_PREV_MON + THU_FRI_TO_NEXT_MON + SAT_SUN_TO_NEXT_MON
+            )
+
+        dec_4 = self._add_holiday_dec_4(
+            # Anniversary of the Coronation of HM King George Tupou I.
+            tr("'Aho Fakamanatu 'o e Hilifaki Kalauni 'o 'Ene 'Afio ko Siaosi Tupou I")
+        )
+        if self._year <= 2009:
+            self._add_observed(dec_4)
+        else:
+            self._add_observed(
+                dec_4, rule=TUE_WED_TO_PREV_MON + THU_FRI_TO_NEXT_MON + SAT_SUN_TO_NEXT_MON
+            )
+
+        # Christmas Day.
+        self._add_christmas_day(tr("'Aho Kilisimasi"))
+
+        # Boxing Day.
+        dec_26 = self._add_christmas_day_two(tr("'Aho 2 'o e Kilisimasi"))
+        if self._year <= 2009:
+            self._add_observed(dec_26, rule=MON_TO_NEXT_TUE)
+
+
+class TO(Tonga):
+    pass
+
+
+class TON(Tonga):
+    pass
+
+
+class TongaStaticHolidays:
+    """
+    References:
+      - https://www.stuff.co.nz/sport/league/99338959/tonga-government-declares-public-holiday-over-rugby-league-teams-deeds  # noqa: E501
+      - https://www.rnz.co.nz/international/pacific-news/398653/tonga-declares-public-holiday-for-pm-pohiva-s-state-funeral  # noqa: E501
+      - https://www.nrl.com/news/2019/11/06/tonga-declare-public-holiday-for-woolfs-winning-warriors/  # noqa: E501
+    """
+
+    # Special Cases.
+
+    # Tonga Rugby Public Holiday.
+    rugby_special_holidays = tr("'Aho malolo 'akapulu 'a Tonga")
+
+    special_public_holidays = {
+        2017: (NOV, 29, rugby_special_holidays),
+        2019: (
+            # State Funeral of 'Akilisi Pohiva.
+            (SEP, 19, tr("Me'afaka'eiki 'o e Siteiti 'Akilisi Pohiva")),
+            (NOV, 15, rugby_special_holidays),
+        ),
+        # Boxing Day (Observed).
+        2021: (DEC, 27, tr("'Aho 2 'o e Kilisimasi (Fakatokanga'i)")),
```

Comment:
Seems we need `special_holidays_observed` :)

Reply:
Seems like things got changed around a bit from [the initial 2021 Public Holidays announcement in late 2020](https://www.gov.to/press-release/tonga-public-holidays-for-2021/) as the additional Dec 26 `Boxing Day (Observed)` [was later retracted as of Dec 14, 2021](https://matangitonga.to/2021/12/14/tonga-no-extra-days-christmas-and-new-year-holidays).
>Sorry folks, in Tonga there will be no extra days off this holiday season to commemorate Christmas Day and New Year's Day because the two public holidays fall on weekends.
>
>The Government of Tonga confirmed that Christmas and New Year will be observed on the actual days they fall on, with no extra day allowed to observe the traditional public holidays.
>
>This year Christmas will be observed on Saturday, December 25 and New Year's Day, which falls on Saturday, 1 January, 2022, will be observed on that day.
>
>This is unlike international holidays in New Zealand and Australia. They get extra days to observe their Christmas and New Year holidays, when it falls during the weekend. This year, both countries will observe Christmas Day on Monday, December 27 and Boxing Day on Tuesday December 28, while New Year's Day will be observed on Monday January 3, 2022.
>
>In Tonga Boxing Day will be a public holiday on Monday, December 27, but there's no extra day for Christmas.
>
>Some private businesses in Tonga, however, will unofficially allow the seasonal holidays for their staff on the Monday and Tuesday

It still wouldn't hurt having that function, though I'm still afraid of ever getting to implement a `special_[subdiv]_[category]_holidays_observed` anytime soon 😭

Reply:
#1535 is ready. Without subdivs and categories yet :)

Reply:
Fixed.